src/support/generation.c

```diff
@@ -0,0 +1,360 @@
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_SESSION_IMPL *s;
+	uint64_t v;
+	uint32_t i, session_cnt;
+	int pause_cnt;
+
+	conn = S2C(session);
+
+	/*
+	 * No lock is required because the session array is fixed size, but it
+	 * may contain inactive entries. We must review any active session, so
+	 * insert a read barrier after reading the active session count. That
+	 * way, no matter what sessions come or go, we'll check the slots for
+	 * all of the sessions that could have been active when we started our
+	 * check.
+	 */
+	WT_ORDERED_READ(session_cnt, conn->session_cnt);
+	for (pause_cnt = 0,
+	    s = conn->sessions, i = 0; i < session_cnt; ++s, ++i) {
+		if (!s->active)
+			continue;
+
+		for (;;) {
+			/* Ensure we only read the value once. */
+			WT_ORDERED_READ(v, s->generations[which]);
+
+			if (v == 0 || generation <= v)
+				break;
+
+			/*
+			 * The pause count is cumulative, quit spinning if it's
+			 * not doing us any good, that can happen in generations
+			 * that don't move quickly.
+			 */
+			if (++pause_cnt < WT_THOUSAND)
+				WT_PAUSE();
+			else
+				__wt_sleep(0, 10);
+		}
+	}
+}
+#endif
+
+/*
+ * __wt_gen_oldest --
+ *	Return the oldest generation in use for the resource.
+ */
+uint64_t
+__wt_gen_oldest(WT_SESSION_IMPL *session, int which)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_SESSION_IMPL *s;
+	uint64_t oldest, v;
+	uint32_t i, session_cnt;
+
+	conn = S2C(session);
+
+	/*
+	 * No lock is required because the session array is fixed size, but it
+	 * may contain inactive entries. We must review any active session, so
+	 * insert a read barrier after reading the active session count. That
+	 * way, no matter what sessions come or go, we'll check the slots for
+	 * all of the sessions that could have been active when we started our
+	 * check.
+	 */
+	WT_ORDERED_READ(session_cnt, conn->session_cnt);
+	for (oldest = conn->generations[which] + 1,
+	    s = conn->sessions, i = 0; i < session_cnt; ++s, ++i) {
+		if (!s->active)
+			continue;
+
+		/* Ensure we only read the value once. */
+		WT_ORDERED_READ(v, s->generations[which]);
+		if (v != 0 && v < oldest)
+			oldest = v;
+	}
+
+	return (oldest);
+}
+
+/*
+ * __wt_session_gen --
+ *	Return the thread's resource generation.
+ */
+uint64_t
+__wt_session_gen(WT_SESSION_IMPL *session, int which)
+{
+	return (session->generations[which]);
+}
+
+#if 0
+/*
+ * __wt_session_gen_publish_next --
+ *	Switch the resource to a new generation, then publish a thread's
+ * resource generation.
+ */
+uint64_t
+	TABBED IN to avoid dist/ functions:
+	__wt_session_gen_publish_next(WT_SESSION_IMPL *session, int which)
+{
+	session->generations[which] = __wt_gen_next(session, which);
+
+	/* Ensure threads waiting on a resource to drain see the new value. */
+	WT_FULL_BARRIER();
+
+	return (session->generations[which]);
+}
+#endif
+
+/*
+ * __wt_session_gen_publish --
+ *	Publish a thread's resource generation.
+ */
+uint64_t
+__wt_session_gen_publish(WT_SESSION_IMPL *session, int which)
```

Comment:
I'd call this `__wt_gen_enter`, and I think it should loop until it sees that the published generation is still current.  I note that the split code is the only thing using it at the moment but I suspect there will be more cases (e.g., eviction).